oat-sa/extension-tao-testqti
# Updated remaining time calculation with timer adjustments

Timer adjustments are included when calculating maximum time limits and remaining time during the test session.
 
Related to : https://oat-sa.atlassian.net/browse/TCA-399
 
`QtiTimeConstraint` maximum remaining time calculation is updated to include adjustments, minimum remaining time calculation is fixed to exclude `extraTime`. Values serialized and provided to FE updated. In order to make sure the consumed extra time calculation is correct the `endItemTimer` was updated. `taoQtiTest_helpers_TestSession::closeTimer()` makes sure that during timeout event the remaining time is 0, updated to include timer adjustments.
 
#### How to test
 
Manual testing can be performed in playground environment created for this feature or local environment with appropriate branches of `taoTests`, `taoQtiTest`, `taoProctoring`. The configuration option `DeliveryExecutionStateService::OPTION_TIME_HANDLING` should be set to `DeliveryExecutionStateService::TIME_HANDLING_TIMER_ADJUSTMENT`.

- As a test taker launch a proctored test with time limits
- As a proctor authorize the test to start, when the test is in progress pause it
- As a proctor refresh the delivery execution list
- As a proctor click on Change time button, specify how many minutes should be added/subtracted and click OK.
- As a proctor authorize the test session
- As a test taker Proceed to the test and see that the timer value is updated
  
#### Dependencies
   
Required:
 - [x] https://github.com/oat-sa/extension-tao-test/pull/323

Companion PR :
 - [ ] https://github.com/oat-sa/extension-tao-proctoring/pull/801
 - [ ] https://github.com/oat-sa/extension-tao-act/pull/1352
